### PR TITLE
OSCI-4011: add ET teiid query for related advisories

### PR DESCRIPTION
**Pull request by Andrei-Stepanov**
Signed-off-by: Andrei Stepanov <astepano@redhat.com>

### CODE REVIEW

**Comment by Andrei-Stepanov** [comment]
This is ready to be merged. @mgrabovsky  please ack.
This code has a new dependency on new `pg` module.
I do not include here change to package-lock.json and package.json
I will create a separete MR for these chages to easier handle MR.

File: src/services/teiid.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+(async function init() {
+  /* Query for testing purpose to check on startup if Teiid is responsive */
+  const testQuery = `
+  SELECT
+        c.id, c.created_at, c.fulladvisory
+  FROM
+        Errata_public.errata_main c
+  WHERE
+	      c.id=(SELECT MAX(id) FROM Errata_public.errata_main)
```
Comment (mgrabovsky) [inline src/services/teiid.ts:64]: There's a tab here that should presumably be a space:
```suggestion
        c.id=(SELECT MAX(id) FROM Errata_public.errata_main)
```

File: src/schema/teiid.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+  GraphQLFieldConfig,
+  GraphQLInt,
+  GraphQLList,
+  GraphQLNonNull,
+  GraphQLObjectType,
+  GraphQLString,
+} from 'graphql';
+import { getTeiidClient } from '../services/teiid';
+
+const log = debug('osci:root_query_type');
```
Comment (mgrabovsky) [inline src/schema/teiid.ts:34]: Should this be adjusted for the module name?
```suggestion
const log = debug('osci:teiid');
```

File: src/schema/teiid.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        c2.errata_id,
+        c3.name,
+        c4.status, c4.fulladvisory
+FROM
+        Errata_public.brew_builds c1,
+        Errata_public.errata_brew_mappings c2,
+        Errata_public.product_versions c3,
+        Errata_public.errata_main c4
+WHERE
+        c1.nvr IN (${qs})
```
Comment (mgrabovsky) [inline src/schema/teiid.ts:82]: This is definitely not secure. You should use a [parameterised query](https://node-postgres.com/features/queries#parameterized-query).

**Comment by Andrei-Stepanov** [comment]
@mgrabovsky it seems doesn't work with TEIID
What I found:
It would work with original PostgreSQL:

https://github.com/brianc/node-postgres/wiki/FAQ#11-how-do-i-build-a-where-foo-in--query-to-find-rows-matching-an-array-of-values

They propose to change `in` for `= any`.

But, TEIID doesn't understand this construction.

Moreover: `Index not used with =any() but used with in` :  https://dba.stackexchange.com/questions/125413/index-not-used-with-any-but-used-with-in/125500#125500

More links:

https://github.com/brianc/node-postgres/issues/2268#issuecomment-822665782
https://stackoverflow.com/questions/34627026/in-vs-any-operator-in-postgresql
https://stackoverflow.com/questions/10720420/node-postgres-how-to-execute-where-col-in-dynamic-value-list-query

For `ANY` TEIID answers:

```
{
  "errors": [
    {
      "message": "TEIID30492 Aggregate functions are only allowed HAVING/SELECT/ORDER BY clauses.  Window functions are only allowed in the SELECT/ORDER BY clauses: [ANY(?)].  Both require a FROM clause to be present.",
      "locations": [
```

Then they recommend to overcome this to construct query manually, what fallback to the original PR.

**Comment by Andrei-Stepanov** [comment]
@mgrabovsky 

Code:  

```ts
/* Put limit just to prevent DoS, with correct arguments there cannot be so much linked errata */
const mkTeiidQueryETLinkedAdvisories = (argsNum: number): string => {
  /* creates a string in form: '$1, $2, $3, ...' */
  const args = _.join(
    _.map(_.range(argsNum), (i) => `$${i + 1}`),
    ', ',
  );
  const query: string = `
SELECT          
        c1.nvr, c1.id,
        c2.errata_id,
        c3.name,
        c4.status, c4.fulladvisory
FROM
        Errata_public.brew_builds c1,
        Errata_public.errata_brew_mappings c2,
        Errata_public.product_versions c3,
        Errata_public.errata_main c4
WHERE
        c1.nvr in (${args})
        AND
        c1.id=c2.brew_build_id
        AND
        c2.product_version_id=c3.id
        AND
        c2.errata_id=c4.id
LIMIT 200
`;
  return query;
};

//..

    const res = await client.query(query, nvrs);
```

Fails in the same way as with params.

```
/Users/astepano/shared/osci/ciboard-server/node_modules/pg/lib/client.js:132
      const error = this._ending ? new Error('Connection terminated') : new Error('Connection terminated unexpectedly')
                                                                        ^
Error: Connection terminated unexpectedly
    at Connection.<anonymous> (/Users/astepano/shared/osci/ciboard-server/node_modules/pg/lib/client.js:132:73)
    at Object.onceWrapper (node:events:627:28)
    at Connection.emit (node:events:513:28)
    at Connection.emit (node:domain:489:12)
    at TLSSocket.<anonymous> (/Users/astepano/shared/osci/ciboard-server/node_modules/pg/lib/connection.js:107:12)
    at TLSSocket.emit (node:events:525:35)
    at TLSSocket.emit (node:domain:489:12)
    at endReadableNT (node:internal/streams/readable:1359:12)
    at processTicksAndRejections (node:internal/process/task_queues:82:21)
```

**Comment by Andrei-Stepanov** [comment]
Escaping just single quote is not enough.

**Comment by Andrei-Stepanov** [comment]
I could not find that TEIID supports arguments, moreover I found:

https://docs.jboss.org/teiid/7.7.0.Final/reference/en-US/html_single/ 

What is called `Native Queries`, that can support argumetns, but, it uses some special format:

`SELECT x.* FROM (call native('....` that is not supported by nodejs postrgresql driver.

It seems that is enough only escape single quote according to: https://www.postgresql.org/docs/current/sql-syntax-lexical.html at 4.1.2.1. String Constants

**Comment by Andrei-Stepanov** [comment]
> Escaping just single quote is not enough.

seems enough

**Comment by Andrei-Stepanov** [comment]
@mgrabovsky I opened a new PR at : https://github.com/fedora-ci/ciboard-server/pull/18
I do not want to do force push here.

**Pull request closed without merging**